paddlenlp/transformers/ctrl/modeling.py

```diff
@@ -0,0 +1,764 @@
+            logits = outputs.logits
+            # [batch_size, vocab_size]
+            logits = logits[:, -1, :]
+
+            # pre-process distribution
+            logits = self.adjust_logits_during_generation(logits)
+            logits = logits_processors(input_ids, logits)
+
+            # beam search
+            # [batch_size * num_beams, vocab_size]
+            next_scores = F.softmax(logits)
+            next_scores = paddle.log(next_scores)
+
+            next_scores = next_scores + beam_scores.unsqueeze(-1)
+            # reshape for beam search
+            vocab_size = next_scores.shape[-1]
+            next_scores = next_scores.reshape(
+                [batch_size, num_beams * vocab_size])
+
+            next_scores, next_tokens = paddle.topk(
+                next_scores, 2 * num_beams, axis=1)
+
+            next_indices = next_tokens // vocab_size
+            next_tokens = next_tokens % vocab_size
+
+            # stateless
+            beam_outputs = beam_scorer.process(
+                input_ids,
+                next_scores,
+                next_tokens,
+                next_indices,
+                origin_len=origin_len,
+                pad_token_id=pad_token_id,
+                eos_token_id=eos_token_id, )
+            beam_scores = beam_outputs["next_beam_scores"]
+            beam_next_tokens = beam_outputs["next_beam_tokens"]
+            beam_idx = beam_outputs["next_beam_indices"]
+
+            cur_len += 1
+            input_ids = paddle.concat(
+                [
+                    paddle.index_select(input_ids, beam_idx),
+                    beam_next_tokens.unsqueeze(-1),
+                ],
+                axis=-1, )
+
+            if beam_scorer.is_done:
+                break
+            model_kwargs = self.update_model_kwargs_for_generation(outputs,
+                                                                   model_kwargs)
+            if model_kwargs["past"] is not None:
+                model_kwargs["past"] = self._reorder_cache(model_kwargs["past"],
+                                                           beam_idx)
+
+        pred_ids, scores = beam_scorer.finalize(
+            input_ids,
+            beam_scores,
+            next_tokens,
+            next_indices,
+            pad_token_id=pad_token_id,
+            eos_token_id=eos_token_id, )
+        return pred_ids[:, origin_len:], scores
+
+    def update_model_kwargs_for_generation(self, outputs, model_kwargs):
+        # update past
+        if "past_key_values" in outputs:
+            model_kwargs["past"] = outputs.past_key_values
+        elif "mems" in outputs:
+            model_kwargs["past"] = outputs.mems
+        elif "past_buckets_states" in outputs:
+            model_kwargs["past"] = outputs.past_buckets_states
+        else:
+            model_kwargs["past"] = None
+
+        return model_kwargs
+
+
+@register_base_model
+class CTRLModel(CTRLPreTrainedModel):
+    def __init__(self, **kwargs):
```

Comment:
Please add detailed documentation for your model. See bert for example

Reply:
done